shared/mempool/generic_fifo_set.go

```diff
@@ -0,0 +1,167 @@
+package mempool
+
+import (
+	"container/list"
+	"fmt"
+	"sync"
+)
+
+type GenericFIFOSet[TIdx comparable, TData comparable] struct {
+	set         map[TIdx]struct{}
+	queue       *list.List
+	maxElements int
+	m           sync.Mutex
+
+	indexerFn     func(any) TIdx
+	isOverflowing func(*GenericFIFOSet[TIdx, TData]) bool
+
+	onAdd       func(TData, *GenericFIFOSet[TIdx, TData])
+	onRemove    func(TData, *GenericFIFOSet[TIdx, TData])
+	onCollision func(TData, *GenericFIFOSet[TIdx, TData]) error
+}
+
+func (g *GenericFIFOSet[TIdx, TData]) Push(item TData) error {
+	g.m.Lock()
+	defer g.m.Unlock()
+
+	index := g.indexerFn(item)
+	if _, ok := g.set[index]; ok {
+		if g.onCollision != nil {
+			return g.onCollision(item, g)
+		}
+		return fmt.Errorf("item %v already exists", item)
+	}
+
+	g.set[index] = struct{}{}
+	g.queue.PushBack(item)
+	g.onAdd(item, g)
+	if g.isOverflowing != nil && g.isOverflowing(g) {
+		//if g.queue.Len() > g.maxElements {
+		front := g.queue.Front()
```

Comment:
Seems like this is somethins that should be the default "customIsOverflowing" but the logic shouldn't be executed if a custom overflow handler is provided

Reply:
It's already implemented in the constructor, I made it more explicit now `defaultIsOverflowing`.